src/esp/physics/URDFImporter.cpp

```diff
@@ -0,0 +1,334 @@
+                                float& jointLowerLimit,
+                                float& jointUpperLimit,
+                                float& jointDamping,
+                                float& jointFriction) const {
+  float jointMaxForce{0.f};
+  float jointMaxVelocity{0.f};
+  return getJointInfo2(linkIndex, parent2joint, linkTransformInWorld,
+                       jointAxisInJointSpace, jointType, jointLowerLimit,
+                       jointUpperLimit, jointDamping, jointFriction,
+                       jointMaxForce, jointMaxVelocity);
+}
+
+void URDFImporter::getMassAndInertia2(int linkIndex,
+                                      float& mass,
+                                      Mn::Vector3& localInertiaDiagonal,
+                                      Mn::Matrix4& inertialFrame) const {
+  if ((flags & CUF_USE_URDF_INERTIA) != 0) {
+    getMassAndInertia(linkIndex, mass, localInertiaDiagonal, inertialFrame);
+  } else {
+    // the link->m_inertia is NOT necessarily aligned with the inertial frame
+    // so an additional transform might need to be computed
+    auto link = activeModel_->getLink(linkIndex);
+    if (link != nullptr) {
+      float linkMass{0.f};
+      if (!link->m_parentJoint.lock() && activeModel_->m_overrideFixedBase) {
+        linkMass = 0.f;
+      } else {
+        linkMass = link->m_inertia.m_mass;
+      }
+      mass = linkMass;
+      localInertiaDiagonal = Mn::Vector3(0, 0, 0);
+      inertialFrame =
+          Mn::Matrix4::from(link->m_inertia.m_linkLocalFrame.rotation(),
+                            link->m_inertia.m_linkLocalFrame.translation());
+    } else {
+      mass = 1.f;
+      localInertiaDiagonal = Mn::Vector3(1, 1, 1);
+      inertialFrame = Mn::Matrix4();  // Identity
+    }
+  }
+}
+
+void URDFImporter::getMassAndInertia(int linkIndex,
+                                     float& mass,
+                                     Mn::Vector3& localInertiaDiagonal,
+                                     Mn::Matrix4& inertialFrame) const {
+  Mn::Debug silence{logMessages ? &std::cout : nullptr};
+  // the link->m_inertia is NOT necessarily aligned with the inertial frame
+  // so an additional transform might need to be computed
+  auto link = activeModel_->getLink(linkIndex);
+  if (link != nullptr) {
+    Mn::Matrix3 linkInertiaBasis;  // Identity
+    float linkMass = 0.f;
+    float principalInertiaX = 0.f;
+    float principalInertiaY = 0.f;
+    float principalInertiaZ = 0.f;
+    if (!link->m_parentJoint.lock() && activeModel_->m_overrideFixedBase) {
+      // no change
+    } else {
+      linkMass = link->m_inertia.m_mass;
+      if (link->m_inertia.m_ixy == 0.0 && link->m_inertia.m_ixz == 0.0 &&
+          link->m_inertia.m_iyz == 0.0) {
+        principalInertiaX = link->m_inertia.m_ixx;
+        principalInertiaY = link->m_inertia.m_iyy;
+        principalInertiaZ = link->m_inertia.m_izz;
+      } else {
+        // by column vector
+        Mn::Matrix3 inertiaTensor(
```

Comment:
I think you can construct with nested BraceInitialization which is cleaner, right @mosra ?

Reply:
See here: https://doc.magnum.graphics/magnum/matrix-vector.html#matrix-vector-construction

Reply:
Double to float conversion would require an explicit cast which is uglier even than this. 🤷

Reply:
It might have been more performance though, right @mosra ?